xmppserver/src/main/java/org/jivesoftware/openfire/pubsub/CachingPubsubPersistenceProvider.java

```diff
@@ -0,0 +1,708 @@
+        operations.add( operation );
+    }
+
+    @Override
+    public void updateSubscription(Node node, NodeSubscription subscription) {
+        log.debug( "Updating node subscription: {} {}", node.getUniqueIdentifier(), subscription.getID() );
+        final List<NodeOperation> operations = nodesToProcess.computeIfAbsent( node.getUniqueIdentifier(), id -> new ArrayList<>() );
+
+        // This subscription update can replace any pending updates of the same subscription (since the last create/delete of the node or subscription change of this affiliate to the node).
+        final ListIterator<NodeOperation> iter = operations.listIterator( operations.size() );
+        while ( iter.hasPrevious() ) {
+            final NodeOperation op = iter.previous();
+            if ( op.action.equals( NodeOperation.Action.UPDATE_SUBSCRIPTION ) ) {
+                if ( subscription.getID().equals( op.subscription.getID() ) ) {
+                    iter.remove(); // This is replaced by the update that's being added.
+                }
+            } else {
+                break; // Operations that precede anything other than the last operations that are UPDATE_AFFILIATE shouldn't be replaced.
+            }
+        }
+
+        final NodeOperation operation = NodeOperation.updateSubscription( node, subscription );
+        operations.add( operation );
+    }
+
+    @Override
+    public void removeSubscription(NodeSubscription subscription) {
+        log.debug( "Removing node subscription: {} {}", subscription.getNode().getUniqueIdentifier(), subscription.getID() );
+
+        final List<NodeOperation> operations = nodesToProcess.computeIfAbsent( subscription.getNode().getUniqueIdentifier(), id -> new ArrayList<>() );
+
+        // This subscription removal can replace any pending creation, update or delete of the same subscription (since the last create/delete of the node or subscription change of this subscription to the node).
+        final ListIterator<NodeOperation> iter = operations.listIterator( operations.size() );
+        while ( iter.hasPrevious() ) {
+            final NodeOperation operation = iter.previous();
+            if ( Arrays.asList( NodeOperation.Action.CREATE_SUBSCRIPTION, NodeOperation.Action.UPDATE_SUBSCRIPTION, NodeOperation.Action.REMOVE_SUBSCRIPTION ).contains( operation.action ) ) {
+                if ( subscription.getID().equals( operation.subscription.getID() ) ) {
+                    iter.remove(); // This is replaced by the update that's being added.
+                }
+            } else {
+                break; // Operations that precede anything other than the last operations that are subscription changes shouldn't be replaced.
+            }
+        }
+        operations.add( NodeOperation.removeSubscription( subscription.getNode(), subscription ) );
+    }
+
+    private void process( final NodeOperation operation ) {
+        switch ( operation.action )
```

Comment:
Again, personal preference, but 
a) would make NodeOperation a param to each of the createNode/updateNode/etc/etc/ methods, and
b) Use a map of operation.action to Consumer<NodeOperation> to turn this method in to a couple of lines (albeit at the cost of more effort in class initialisation.

Reply:
The methods that are used now were largely pre-existing, and used by others. We could either overload them, or refactor the lot, but I don't think that that effort outweighs the benefit.